docs/typescript.md

```diff
@@ -2,13 +2,504 @@
+
+##### Redundant or unnecessary `as` assertions are not flagged for removal
+
+Type assertions can also cause false positives, because they assertions are independent expressions, untied to the type errors they were intended to fix. Thus, even if code drift fixes or removes a particular type error, the type assertions that were put in place to fix that error will provide no indication that they are no longer necessary and now should be removed.
+
+```typescript
+enum Direction {
+  Up = 'up',
+  Down = 'down',
+  Left = 'left',
+  Right = 'right',
+}
+const directions = Object.values(Direction);
+
+// Error: Element implicitly has an 'any' type because index expression is not of type 'number'.(7015)
+// Only one of the two `as` assertions necessary to fix error, but neither are flagged as redundant.
+for (const key of Object.keys(directions) as keyof directions[]) {
+  const direction = directions[key as keyof typeof directions];
+}
+```
+
+##### Type guards can be used to improve type inference
+
+```typescript
+function isSomeInterface(x: unknown): x is SomeInterface {
+  return (
+    'name' in x &&
+    typeof x.name === 'string' &&
+    'length' in x &&
+    typeof x.length === 'number'
+  );
+}
+```
+
+🚫 Type assertion
+
+```typescript
+function f(x: SomeInterface | SomeOtherInterface) {
+  console.log((x as SomeInterface).name);
+}
+```
+
+✅ Narrowing with type guard
+
+```typescript
+function f(x: SomeInterface | SomeOtherInterface) {
+  if (isSomeInterface(x)) {
+    // Type of x: 'SomeInterface | SomeOtherInterface'
+    console.log(x.name); // Type of x: 'SomeInterface'. Type of x.name: 'string'.
+  }
+}
+```
+
+```typescript
+const nftMetadataResults = await Promise.allSettled(...);
+
+nftMetadataResults
+  .filter((promise) => promise.status === 'fulfilled')
+  .forEach((elm) =>
+    this.updateNft(
+      elm.value.nft, // Property 'value' does not exist on type 'PromiseRejectedResult'.ts(2339)
+      ...
+    ),
+  );
+```
+
+🚫 Type assertion
+
+```typescript
+(nftMetadataResults.filter(
+    (promise) => promise.status === 'fulfilled',
+  ) as { status: 'fulfilled'; value: NftUpdate }[])
+  .forEach((elm) =>
+    this.updateNft(
+      elm.value.nft,
+      ...
+    ),
+  );
+```
+
+✅ Use a type guard as the predicate for the filter operation, enabling TypeScript to narrow the filtered results to `PromiseFulfilledResult` at the type level
+
+```typescript
+nftMetadataResults.filter(
+    (result): result is PromiseFulfilledResult<NftUpdate> =>
+      result.status === 'fulfilled',
+  )
+  .forEach((elm) =>
+    this.updateNft(
+      elm.value.nft,
+      ...
+    ),
+  );
+```
+
+#### Acceptable usages of `as`
```

Comment:
C: we often have to use as to type a process.env value that is always considered a string. Maybe this could be added as example?

Reply:
Absolutely! This sounds like a great example for an externally determined type. Would you be able to provide an example for this?